### PR TITLE
TS Delivery SDK v4 release

### PR DESCRIPTION
**Pull request by nadeem-cs**
Changes for JavaScript to Typescript

### CODE REVIEW

File: test/utils/constant.ts [hunk 1]
```diff
@@ -0,0 +1,6 @@
+export const HOST_URL = 'cdn.contentstack.io';
+export const LOCALE = 'en-155';
+export const HOST_EU_REGION = 'eu-cdn.contentstack.com';
+export const HOST_AZURE_NA_REGION = 'azure-na-cdn.contentstack.com';
```
Comment (sunil-lakshman) [inline test/utils/constant.ts:4]: @nadeem-cs add azure_eu region as well, 
URL - "azure-eu-cdn.contentstack.com"

File: test/api/asset.spec.ts [hunk 2]
```diff
@@ -0,0 +1,92 @@
+/* eslint-disable no-console */
+/* eslint-disable promise/always-return */
+import { Asset } from '../../src/lib/asset';
+import { stackInstance } from '../utils/stack-instance';
+import { TAsset } from './types';
+
+const stack = stackInstance();
+const assetUid = 'blt122ea52c5d4ddf19';
```
Comment (ishaileshmishra) [inline test/api/asset.spec.ts:8]: Please remove assetUid from the asset.apec.ts file @nadeem-cs

File: test/api/contenttype.spec.ts [hunk 3]
```diff
@@ -0,0 +1,29 @@
+/* eslint-disable no-console */
+/* eslint-disable promise/always-return */
+import { ContentType } from '../../src/lib/content-type';
+import { stackInstance } from '../utils/stack-instance';
+import { TContentType, TEntry } from './types';
+
+const stack = stackInstance();
+describe('ContentType API test cases', () => {
+  it('should give Entry instance when entry method is called with entryUid', async () => {
+    const result = await makeContentType('header').Entry('blt657fe7db362fea22').fetch<TEntry>();
```
Comment (ishaileshmishra) [inline test/api/contenttype.spec.ts:10]: remove the credentials from the file @nadeem-cs

File: test/api/entry.spec.ts [hunk 4]
```diff
@@ -0,0 +1,59 @@
+import { BaseEntry } from 'src';
+import { Entry } from '../../src/lib/entry';
+import { stackInstance } from '../utils/stack-instance';
+import { TEntry } from './types';
+
+const stack = stackInstance();
+
+describe('Entry API tests', () => {
+  it('should check for entry is defined', async () => {
+    const result = await makeEntry('blt09f7d2d46afe6dc6').fetch<TEntry>();
```
Comment (ishaileshmishra) [inline test/api/entry.spec.ts:10]: remove the credentials from the file @nadeem-cs

File: test/api/entry.spec.ts [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+    expect(result.entry.created_by).toBeDefined();
+    expect(result.entry.updated_by).toBeDefined();
+  });
+
+  it('should check for entry is defined with BaseEntry', async () => {
+    interface MyEntry extends BaseEntry {
+      bio: string;
+      age: string;
+    }
+    const result = await makeEntry('blt09f7d2d46afe6dc6').fetch<MyEntry>();
```
Comment (ishaileshmishra) [inline test/api/entry.spec.ts:24]: remove the credentials from the file @nadeem-cs

File: test/api/entry.spec.ts [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+  });
+  it('should check for include branch', async () => {
+    const result = await makeEntry('blt09f7d2d46afe6dc6').includeBranch().fetch<TEntry>();
+    expect(result.entry._branch).not.toEqual(undefined);
+    expect(result.entry.uid).toBeDefined();
+    expect(result.entry.created_by).toBeDefined();
+    expect(result.entry.updated_by).toBeDefined();
+  });
+  it('should check for locale', async () => {
+    const result = await makeEntry('blt09f7d2d46afe6dc6').locale('fr-fr').fetch<TEntry>();
```
Comment (ishaileshmishra) [inline test/api/entry.spec.ts:37]: remove the credentials from the file @nadeem-cs

File: test/api/entry.spec.ts [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+      bio: string;
+      age: string;
+    }
+    const result = await makeEntry('blt09f7d2d46afe6dc6').fetch<MyEntry>();
+
+    console.log('🚀 ~ file: entry.spec.ts:25 ~ it ~ result:', result);
+    expect(result.entry).toBeDefined();
+  });
+  it('should check for include branch', async () => {
+    const result = await makeEntry('blt09f7d2d46afe6dc6').includeBranch().fetch<TEntry>();
```
Comment (ishaileshmishra) [inline test/api/entry.spec.ts:30]: remove the credentials from the file @nadeem-cs

File: test/api/entry.spec.ts [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+    const result = await makeEntry('blt09f7d2d46afe6dc6').locale('fr-fr').fetch<TEntry>();
+    expect(result.entry).toBeDefined();
+    expect(result.entry._version).toBeDefined();
+    expect(result.entry.publish_details.locale).toEqual('fr-fr');
+    expect(result.entry.uid).toBeDefined();
+    expect(result.entry.created_by).toBeDefined();
+    expect(result.entry.updated_by).toBeDefined();
+  });
+  it('should check for include fallback', async () => {
+    const result = await makeEntry('blt09f7d2d46afe6dc6').includeFallback().fetch<TEntry>();
```
Comment (ishaileshmishra) [inline test/api/entry.spec.ts:46]: remove the credentials from the file @nadeem-cs

File: test/api/query.spec.ts [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+      .find<TEntries>();
+    expect(query.entries[0].title).toEqual('The future of business with AI');
+  });
+  it('should add a where filter to the query parameters', async () => {
+    const query = await makeQuery('blog_post').where('_version', QueryOperation.IS_LESS_THAN, 3).find<TEntries>();
+    expect(query.entries[0].title).toEqual('The future of business with AI');
+  });
+  it('should add a where-in filter to the query parameters', async () => {
+    const query = await makeQuery('blog_post')
+      .whereIn('author', makeQuery('author').where('uid', QueryOperation.EQUALS, 'blt09f7d2d46afe6dc6'))
```
Comment (ishaileshmishra) [inline test/api/query.spec.ts:19]: remove the credentials from the file @nadeem-cs

File: test/api/query.spec.ts [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+      .find<TEntries>();
+    expect(query.entries[0].author[0].uid).toEqual('blt09f7d2d46afe6dc6');
+    expect(query.entries[0].title).toBeDefined();
+    expect(query.entries[0].url).toBeDefined();
+    expect(query.entries[0]._version).toBeDefined();
+    expect(query.entries[0].publish_details).toBeDefined();
+  });
+  it('should add a whereNotIn filter to the query parameters', async () => {
+    const query = await makeQuery('blog_post')
+      .whereNotIn('author', makeQuery('author').where('uid', QueryOperation.EQUALS, 'blt09f7d2d46afe6dc6'))
```
Comment (ishaileshmishra) [inline test/api/query.spec.ts:29]: remove the credentials from the file @nadeem-cs

File: test/api/query.spec.ts [hunk 11]
```diff
@@ -0,0 +1,53 @@
...
+      .find<TEntries>();
+    expect(query.entries[0].author[0]).not.toEqual('blt09f7d2d46afe6dc6');
+    expect(query.entries[0].title).toBeDefined();
+    expect(query.entries[0].url).toBeDefined();
+    expect(query.entries[0]._version).toBeDefined();
+    expect(query.entries[0].publish_details).toBeDefined();
+  });
+  it('should add a query operator to the query parameters', async () => {
+    const query1 = makeQuery('blog_post').where('locale', QueryOperation.EQUALS, 'en-us');
+    const query2 = makeQuery('blog_post').where('created_by', QueryOperation.EQUALS, 'blt79e6de1c5230a991');
```
Comment (ishaileshmishra) [inline test/api/query.spec.ts:39]: remove the credentials from the file @nadeem-cs

File: test/unit/image-transform.spec.ts [hunk 12]
```diff
@@ -0,0 +1,253 @@
...
+    expect(getBuild(new ImageTransform().trim(50))).toEqual({ trim: '50' });
+  });
+
+  it('should return valid object when orient method is called with valid params', () => {
+    const imgTObj = getBuild(new ImageTransform().orient(Orientation.FLIP_HORIZONTAL));
+    expect(imgTObj).toEqual({ orient: '2' });
+  });
+
+  it('should return valid object when overlay method is called with valid params', () => {
+    const overlayImgURL = '/v3/assets/blteae40eb499811073/bltb21dacdd20d0e24c/59e0c401462a293417405f34/circle.png';
```
Comment (ishaileshmishra) [inline test/unit/image-transform.spec.ts:124]: remove the credentials from the file @nadeem-cs

**Review by ishaileshmishra — CHANGES_REQUESTED** [review]
found multiple credentials. Pls remove the credentials from the file @nadeem-cs

File: test/api/asset.spec.ts (hunk 2 above)
Comment (nadeem-cs) [reply test/api/asset.spec.ts]: Resolved in [this](https://github.com/contentstack/contentstack-javascript/pull/110) PR.

File: test/api/contenttype.spec.ts (hunk 3 above)
Comment (nadeem-cs) [reply test/api/contenttype.spec.ts]: Resolved in [this](https://github.com/contentstack/contentstack-javascript/pull/110) PR.

File: test/api/entry.spec.ts (hunk 4 above)
Comment (nadeem-cs) [reply test/api/entry.spec.ts]: Resolved in [this](https://github.com/contentstack/contentstack-javascript/pull/110) PR.

File: test/api/entry.spec.ts (hunk 6 above)
Comment (nadeem-cs) [reply test/api/entry.spec.ts]: Resolved in [this](https://github.com/contentstack/contentstack-javascript/pull/110) PR.

File: test/api/entry.spec.ts (hunk 7 above)
Comment (nadeem-cs) [reply test/api/entry.spec.ts]: Resolved in [this](https://github.com/contentstack/contentstack-javascript/pull/110) PR.

File: test/api/query.spec.ts (hunk 9 above)
Comment (nadeem-cs) [reply test/api/query.spec.ts]: Resolved in [this](https://github.com/contentstack/contentstack-javascript/pull/110) PR.

File: test/api/query.spec.ts (hunk 11 above)
Comment (nadeem-cs) [reply test/api/query.spec.ts]: Resolved in [this](https://github.com/contentstack/contentstack-javascript/pull/110) PR.

File: test/unit/image-transform.spec.ts (hunk 12 above)
Comment (nadeem-cs) [reply test/unit/image-transform.spec.ts]: Resolved in [this](https://github.com/contentstack/contentstack-javascript/pull/110) PR.

**Comment by nadeem-cs** [comment]
All the Uid key leaks are resolved in [this](https://github.com/contentstack/contentstack-javascript/pull/110) PR.

**Review by ishaileshmishra — APPROVED** [review]
approved. 

- Ensure that all tests pass before merging this PR, given its substantial size and scope 
- review failing checks

**Comment by nadeem-cs** [comment]
Will be releasing the v4 version from [v4-release](https://github.com/contentstack/contentstack-javascript/tree/v4-release) branch and not from master. Hence closing this PR.

**Pull request closed without merging**